### ISSUE TITLE
Further development sup vs. astroid

### ISSUE THREAD

**Issue by maxmeyer** [sup-heliotrope/sup#504]
@gauteh I saw you developing another MUA. Does it makes sense to try that one? 

Can you please eloborate a bit on sup vs. astroid? Is sup required to run `astroid`. If not you may want to change the wording of "A graphical threads-with-tags mail user agent based on sup and notmuch. " a bit.

Will it be a single binary? Does `sup` have a future when `astroid` becomes more mature?

**Comment by binaryplease** [comment]
When talking about/comparing notmuch MUAs maybe alot (https://github.com/pazz/alot)  should be considered too.

**Comment by maxmeyer** [comment]
I think this would be a topic of another thread. Thanks.

**Comment by gauteh** [comment]
Dennis Günnewig writes on January 16, 2016 18:46:
> @gauteh I saw you developing another MUA. Does it makes sense to try that one? 

Yes. Especially if you are experimenting with notmuch, testing out
astroid is hopefully not much extra hassle. astroid is a newer project
and not yet as mature as sup, features, bugs and tweaks are constantly
being added though, with the good help from a few other programmers and
users.
 
> Can you please eloborate a bit on sup vs. astroid? Is sup required to
> run `astroid`. If not you may want to change the wording of "A
> graphical threads-with-tags mail user agent based on sup and notmuch.
> " a bit.

astroid is a completely new project, it's relation to sup is historical.
notmuch was developed as an answer to sup, and it implementes the mail
database (the indexing - not the content). this allows the user
interface to be nicely separated from all the indexing/db stuff. Indeed,
this is the main reason I have moved away from sup, the overarching
issues with the design of sup would require separation of indexing and
ui, better ui-threading, better mail parsing, etc. to do that would be
to re-invent notmuch.

another thing that I missed in sup was: display of html, attachment
preview, easier handling of unicode and text, ui library that handles
all the threading/updating stuff. as well as the possibility to have
several instances or windows of the mail client open at the same time
(stuff that is implemented and working in astroid now).

[btw: I noticed I used 'based on' in the readme, but what I meant is
'inspired by']
 
> Will it be a single binary? Does `sup` have a future when `astroid`
> becomes more mature?

first; I do not use sup day-to-day anymore. for me, I regrettably do not
think sup has a future - but I did not want to abandon it completely
without anybody else stepping up. as it seems now, there still are core,
dedicated, users, but not necessarily with the time to maintain and fix
things.

astroid is one binary, but it relies on a few data files as well as a
few configuration files. installed it typically consists of the
following files:

```
/usr/bin/astroid
/usr/share/astroid/ui/thread-view.html
/usr/share/astroid/ui/thread-view.scss
/usr/share/astroid/ui/no-mail.png
/usr/share/astroid/ui/edit-message.glade
/usr/share/astroid/ui/acknowledgement.glade
```

minimal configuration consists of:
```
~/.config/astroid/config
```

in addition to notmuch stuff, which you set up manually [0].

you can also configure keybindings and the thread-view (which is HTML
and SCSS (CSS) based).

There is a tour of how to use astroid in the wiki: https://github.com/gauteh/astroid/wiki

Regards, Gaute

[0] https://github.com/gauteh/astroid/wiki/Astroid-in-your-general-mail-setup

**Comment by maxmeyer** [comment]
Hi Gaute,

thanks a lot for your answer. This will become a bit of topic - at least 
for the sup project.

On 17/01/16 10:15, Gaute Hope wrote:
> features, bugs and tweaks are constantly
:)

>
> > Can you please eloborate a bit on sup vs. astroid? Is sup required to
> > run `astroid`. If not you may want to change the wording of "A
> > graphical threads-with-tags mail user agent based on sup and notmuch.
> > " a bit.
>
> astroid is a completely new project, it's relation to sup is historical.
> notmuch was developed as an answer to sup, and it implementes the mail
> database (the indexing - not the content). this allows the user
> interface to be nicely separated from all the indexing/db stuff. Indeed,
> this is the main reason I have moved away from sup, the overarching
> issues with the design of sup would require separation of indexing and
> ui, better ui-threading, better mail parsing, etc. to do that would be
> to re-invent notmuch.

Yeah, yesterday night I made the move of the maildir to dovecot + 
maildir. And just use Thunderbird now as "frontend". If I understand 
"notmuch"'s role correctly, it's not suitable for such a setup, correct?

Although I find your approach with "astroid" quite interesting. I love 
the one binary approach. No hassle with missing gems after a ruby 
update. :-(

 > several instances or windows of the mail client open at the same time

That's something I missed most.

 > ~~~
 > /usr/share/astroid/ui/thread-view.html
 > /usr/share/astroid/ui/thread-view.scss
 > ~~~

You render stuff with HTML+CSS? Cool!


 > There is a tour of how to use astroid in the wiki: 
https://github.com/gauteh/astroid/wiki

Looks promising. I only miss a good frontend for managing contacts. I 
had a look a notmuch-abook. Since I use/used (need to migrate that into 
thunderbird) abook, that would be totally ok for me, but

* Python bindings.... Not good, you got rid of ruby, so please get rid 
of python as well ;-)
* CardDAV support missing -> I set up thunderbird to sync with my 
owncloud I set up at home, works great. This way I've got my android 
phone synced with my local MUA

I'm not sure, how/why notmuch needs a binding with abook: astroid starts 
the new mail, right?

Cheers and a fine snowy weekend (at least here in Germany snow is 
everywhere)

Dennis

**Comment by gauteh** [comment]
Dennis Günnewig writes on January 17, 2016 11:07:
> Yeah, yesterday night I made the move of the maildir to dovecot + 
> maildir. And just use Thunderbird now as "frontend". If I understand 
> "notmuch"'s role correctly, it's not suitable for such a setup, correct?

I think notmuch can work with dovecot (that is my impression from a bit
of googling), which is really the limitation here. astroid works with
anything notmuch works with.

astroid is just the frontend (MUA).
 
> Looks promising. I only miss a good frontend for managing contacts. I 
> had a look a notmuch-abook. Since I use/used (need to migrate that into 
> thunderbird) abook, that would be totally ok for me, but
> 
> * Python bindings.... Not good, you got rid of ruby, so please get rid 
> of python as well ;-)

There are no python hooks yet. No hooks at all actually, the first plan is
to just use custom hooks with shell scripts [0]. notmuch can be manipulated
from the command line, so you could accomplish most things using just
the notmuch cli tool.

> * CardDAV support missing -> I set up thunderbird to sync with my 
> owncloud I set up at home, works great. This way I've got my android 
> phone synced with my local MUA
> 
> I'm not sure, how/why notmuch needs a binding with abook: astroid starts 
> the new mail, right?

astroid doesn't care about contacts. But I use gvim as the editor.
notmuch-abook is used as a plugin in gvim to get contacts from notmuch,
you could really use anything that works with gvim.
 
> Cheers and a fine snowy weekend (at least here in Germany snow is 
> everywhere)

Yes, here in Norway too!

[0] https://github.com/gauteh/astroid/issues/29

**Comment by maxmeyer** [comment]
>
> > Looks promising. I only miss a good frontend for managing contacts. I
> > had a look a notmuch-abook. Since I use/used (need to migrate that into
> > thunderbird) abook, that would be totally ok for me, but
> >
> > * Python bindings.... Not good, you got rid of ruby, so please get rid
> > of python as well ;-)
>
> There are no python hooks yet. No hooks at all actually, the first plan is
> to just use custom hooks with shell scripts [0]. notmuch can be 
> manipulated
> from the command line, so you could accomplish most things using just
> the notmuch cli tool.

No, sorry I was not that clear about what I meant. I think, 
notmuch-abook is implemented in Python - at least is this what github 
language stats told me (https://github.com/guyzmo/notmuch-abook). As I 
like your approach of a single binary, I asked not to rely on another 
interpreter for adding contact list support.
>
> astroid doesn't care about contacts. But I use gvim as the editor.
> notmuch-abook is used as a plugin in gvim to get contacts from notmuch,
> you could really use anything that works with gvim.
Ah. Ok. So it's the same like in sup? astroid creates a temporary file 
which can be parsed by msmtp? That file is opened in $EDITOR - vim or 
whatever is configured. The user then needs to setup a "binding" to 
his/her contact list manager to get contacts "into" the opened file. 
After that information has been added, the user can write his/her mail 
and send it to the recipient by closing the window and pressing "y" or 
whatever is configured. Correct?

**Comment by gauteh** [comment]
Dennis Günnewig writes on January 17, 2016 20:11:
> Ah. Ok. So it's the same like in sup? astroid creates a temporary file 
> which can be parsed by msmtp? That file is opened in $EDITOR - vim or 
> whatever is configured. The user then needs to setup a "binding" to 
> his/her contact list manager to get contacts "into" the opened file. 
> After that information has been added, the user can write his/her mail 
> and send it to the recipient by closing the window and pressing "y" or 
> whatever is configured. Correct?

Yes.

The editor is embedded in the window, the final email is preview in
astroid, you can have a look at the screenshots. Currently I have only
tested with gvim.

**Comment by quark-zju** [comment]
I think Ruby is a productive tool that interests me to sup. With ActiveRecord I could complete a "Patchwork" database and draw them in the UI in relatively short time. It's just very flexible.

Separating the index from ui seems to be a better architecture. It allows a more flexible user-case - like having different sup binaries for different type of buffers and run them in parallel - tiling wm users may love that.

I have tried DRb to separate the index object but failed. I then tried to make sup a frontend of notmuch - it's still [an ongoing work](https://github.com/quark-zju/sup/tree/notmuch) but seems a bit promising.
